File: ObjCLoadWebviewImage/ObjCLoadWebviewImage/ViewController.m
//
//  ViewController.m
//  ObjCLoadWebviewImage
//
//  Created by huangyibiao on 15/10/16.
//  Copyright © 2015年 huangyibiao. All rights reserved.
//

#import "ViewController.h"
#import <CommonCrypto/CommonDigest.h>
#import <AFNetworking/AFNetworking.h>
#import "UIImageView+AFNetworking.h"

@interface ViewController ()

@property (nonatomic, strong) UIWebView *webView;
@property (nonatomic, strong) NSMutableArray *imageViews;

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  self.webView = [[UIWebView alloc] initWithFrame:self.view.bounds];
  self.webView.scalesPageToFit = YES;
  [self.view addSubview:self.webView];
  
  NSURL *url = [[NSBundle mainBundle] URLForResource:@"test" withExtension:@"html"];
  NSString *html = [[NSString alloc] initWithContentsOfURL:url encoding:NSUTF8StringEncoding error:nil];
  
  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"<img\\ssrc[^>]*/>" options:NSRegularExpressionAllowCommentsAndWhitespace error:nil];
  NSArray *result = [regex matchesInString:html options:NSMatchingReportCompletion range:NSMakeRange(0, html.length)];
  
  NSMutableDictionary *urlDicts = [[NSMutableDictionary alloc] init];
  NSString *docPath = [NSHomeDirectory() stringByAppendingPathComponent:@"Documents"];
  
  for (NSTextCheckingResult *item in result) {
    NSString *imgHtml = [html substringWithRange:[item rangeAtIndex:0]];
    
    NSArray *tmpArray = nil;
    if ([imgHtml rangeOfString:@"src=\""].location != NSNotFound) {
      tmpArray = [imgHtml componentsSeparatedByString:@"src=\""];
    } else if ([imgHtml rangeOfString:@"src="].location != NSNotFound) {
      tmpArray = [imgHtml componentsSeparatedByString:@"src="];
    }
    
    if (tmpArray.count >= 2) {
      NSString *src = tmpArray[1];
      
      NSUInteger loc = [src rangeOfString:@"\""].location;
      if (loc != NSNotFound) {
        src = [src substringToIndex:loc];
        
        NSLog(@"正确解析出来的SRC为：%@", src);
        if (src.length > 0) {
          NSString *localPath = [docPath stringByAppendingPathComponent:[self md5:src]];
          // 先将链接取个本地名字，且获取完整路径
          [urlDicts setObject:localPath forKey:src];
        }
      }
    }
  }
  
  // 遍历所有的URL，替换成本地的URL，并异步获取图片
  for (NSString *src in urlDicts.allKeys) {
    NSString *localPath = [urlDicts objectForKey:src];
    html = [html stringByReplacingOccurrencesOfString:src withString:localPath];
    
    // 如果已经缓存过，就不需要重复加载了。
    if (![[NSFileManager defaultManager] fileExistsAtPath:localPath]) {
      [self downloadImageWithUrl:src];
    }
  }
  
  NSLog(@"%@", html);
  
  [self.webView loadHTMLString:html baseURL:url];
}

- (void)downloadImageWithUrl:(NSString *)src {
  // 注意：这里并没有写专门下载图片的代码，就直接使用了AFN的扩展，只是为了省麻烦而已。
  UIImageView *imgView = [[UIImageView alloc] init];
  
  NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:src]];
  [imgView setImageWithURLRequest:request placeholderImage:nil success:^(NSURLRequest * _Nonnull request, NSHTTPURLResponse * _Nullable response, UIImage * _Nonnull image) {
    NSData *data = UIImagePNGRepresentation(image);
    NSString *docPath = [NSHomeDirectory() stringByAppendingPathComponent:@"Documents"];
    NSString *localPath = [docPath stringByAppendingPathComponent:[self md5:src]];
    
    if (![data writeToFile:localPath atomically:NO]) {
      NSLog(@"写入本地失败：%@", src);
    }
  } failure:^(NSURLRequest * _Nonnull request, NSHTTPURLResponse * _Nullable response, NSError * _Nonnull error) {
    NSLog(@"download image url fail: %@", src);
  }];
  
  if (self.imageViews == nil) {
    self.imageViews = [[NSMutableArray alloc] init];
  }
  [self.imageViews addObject:imgView];
}

- (NSString *)md5:(NSString *)sourceContent {
  if (self == nil || [sourceContent length] == 0) {
    return nil;
  }
  
  unsigned char digest[CC_MD5_DIGEST_LENGTH], i;
  CC_MD5([sourceContent UTF8String], (int)[sourceContent lengthOfBytesUsingEncoding:NSUTF8StringEncoding], digest);
  NSMutableString *ms = [NSMutableString string];
  
  for (i = 0; i < CC_MD5_DIGEST_LENGTH; i++) {
    [ms appendFormat:@"%02x", (int)(digest[i])];
  }
  
  return [ms copy];
}

@end

File: iOSLoadWebviewImageSwift/iOSLoadWebviewImage/ViewController.swift
//
//  ViewController.swift
//  iOSLoadWebviewImage
//
//  Created by huangyibiao on 15/10/15.
//  Copyright © 2015年 huangyibiao. All rights reserved.
//

import UIKit


extension String {
  var md5 : String{
    let str = self.cStringUsingEncoding(NSUTF8StringEncoding)
    let strLen = CC_LONG(self.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
    let digestLen = Int(CC_MD5_DIGEST_LENGTH)
    let result = UnsafeMutablePointer<CUnsignedChar>.alloc(digestLen);
    
    CC_MD5(str!, strLen, result);
    
    let hash = NSMutableString();
    for i in 0 ..< digestLen {
      hash.appendFormat("%02x", result[i]);
    }
    result.destroy();
    
    return String(format: hash as String)
  }
}

class ViewController: UIViewController {
  var webView = UIWebView()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    webView.frame = self.view.bounds
    self.view.addSubview(webView)
    
    webView.scalesPageToFit = true
    
    let path = NSBundle.mainBundle().pathForResource("test", ofType: "html")
    let url = NSURL(fileURLWithPath: path!)
    
    do {
      let html = try String(contentsOfURL: url, encoding: NSUTF8StringEncoding)
      //      print(html)
      
      // 获取所有img src中的src链接，并将src更改名称
      // 这里直接采用同步获取数据，异步也是一样的道理，为了方便写demo，仅以同步加载图片为例。
      // 另外，这不考虑清除缓存的问题。
      do {
        let regex = try NSRegularExpression(pattern: "<img\\ssrc[^>]*/>", options: .AllowCommentsAndWhitespace)
        
        let result = regex.matchesInString(html, options: .ReportCompletion, range: NSMakeRange(0, html.characters.count))
        
        var content = html as NSString
        var sourceSrcs: [String: String] = ["": ""]
        
        for item in result {
          let range = item.rangeAtIndex(0)
          
          let imgHtml = content.substringWithRange(range) as NSString
          var array = [""]
          
          if imgHtml.rangeOfString("src=\"").location != NSNotFound {
            array = imgHtml.componentsSeparatedByString("src=\"")
          } else if imgHtml.rangeOfString("src=").location != NSNotFound {
            array = imgHtml.componentsSeparatedByString("src=")
          }
          
          if array.count >= 2 {
            var src = array[1] as NSString
            if src.rangeOfString("\"").location != NSNotFound {
              src = src.substringToIndex(src.rangeOfString("\"").location)
              
              // 图片链接正确解析出来
              print(src)
              
              // 加载图片
              // 这里不处理重复加载的问题，实际开发中，应该要做一下处理。
              // 也就是先判断是否已经加载过，且未清理掉该缓存的图片。如果
              // 已经缓存过，否则才执行下面的语句。
              let data = NSData(contentsOfURL: NSURL(string: src as String)!)
              let localUrl = self.saveImageData(data!, name: (src as String).md5)
              
              // 记录下原URL和本地URL
              // 如果用异步加载图片的方式，先可以提交将每个URL起好名字，由于这里使用的是原URL的md5作为名称，
              // 因此每个URL的名字是固定的。
              sourceSrcs[src as String] = localUrl
            }
          }
        }
        
        for (src, localUrl) in sourceSrcs {
          if !localUrl.isEmpty {
            content = content.stringByReplacingOccurrencesOfString(src as String, withString: localUrl, options: NSStringCompareOptions.LiteralSearch, range: NSMakeRange(0, content.length))
          }
        }
        
        print(content as String)
        webView.loadHTMLString(content as String, baseURL: url)
      } catch {
        print("match error")
      }
    } catch {
      print("load html error")
    }
  }
  
func saveImageData(data: NSData, name: String) ->String {
  let docPath = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true)[0] as NSString
  
  let path = docPath.stringByAppendingPathComponent(name)
  
  // 若已经缓存过，就不需要重复操作了
  if NSFileManager.defaultManager().fileExistsAtPath(path) {
    return path
  }
  
  do {
    try data.writeToFile(path, options: NSDataWritingOptions.DataWritingAtomic)
  } catch {
    print("save image data with name: \(name) error")
  }
  
  return path
}

}

